## [作業重點]
使用 Sklearn 中的 Lasso, Ridge 模型，來訓練各種資料集，務必了解送進去模型訓練的**資料型態**為何，也請了解模型中各項參數的意義。

機器學習的模型非常多種，但要訓練的資料多半有固定的格式，確保你了解訓練資料的格式為何，這樣在應用新模型時，就能夠最快的上手開始訓練！

## 練習時間
試著使用 sklearn datasets 的其他資料集 (boston, ...)，來訓練自己的線性迴歸模型，並加上適當的正則話來觀察訓練情形。

In [0]:
import numpy as np
import matplotlib.pyplot as plt

from sklearn import datasets, linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score  #指標 mean_squared_error

In [0]:
#讀取波士頓房價資料集#
boston = datasets.load_boston()

In [6]:
print(boston.DESCR)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

In [8]:
#讀取資料集
X = boston.data
len(X)

506

In [9]:
#讀取目標
Y = boston.target
len(Y)

506

In [0]:
# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size = 0.2,random_state =4)

In [0]:
#建立線性回歸
regr = linear_model.LinearRegression()

#將訓練資料(train)丟進去模型訓練; fit
regr.fit(x_train, y_train)

#predict y
y_pred = regr.predict(x_test)

In [12]:
#取得回歸係數
print(regr.coef_)

[-1.15966452e-01  4.71249231e-02  8.25980146e-03  3.23404531e+00
 -1.66865890e+01  3.88410651e+00 -1.08974442e-02 -1.54129540e+00
  2.93208309e-01 -1.34059383e-02 -9.06296429e-01  8.80823439e-03
 -4.57723846e-01]


In [16]:
print("Mean_squared_error: %.2d" %mean_squared_error(y_test, y_pred))

Mean_squared_error: 25


# **LASSO**

In [0]:
#讀入波士頓資料集
bos = datasets.load_boston()

In [0]:
# 切分訓練集/測試集
x1_train, x1_test, y1_train, y1_test = train_test_split(bos.data, bos.target, test_size = 0.2, random_state = 4)

In [0]:
# 建立一個線性回歸模型
lasso = linear_model.Lasso(alpha = 1.0)

In [0]:
# 將訓練資料丟進去模型訓練
lasso.fit(x1_train, y1_train)

#predict y
y1_pred = lasso.predict(x1_test)

In [21]:
# 印出各特徵對應的係數，可以看到許多係數都變成 0，Lasso Regression 的確可以做特徵選取
lasso.coef_

array([-0.06494981,  0.04581458, -0.        ,  0.        , -0.        ,
        1.18140024,  0.01109101, -0.73695809,  0.23350042, -0.01551065,
       -0.69270805,  0.00763157, -0.6927848 ])

In [22]:
# 預測值與實際值的差距，使用 MSE
print('MSE: %.2f' %mean_squared_error(y1_test, y1_pred))

MSE: 28.95


 LASS 的結果並沒有比原本的線性回歸來得好， 這是因為目標函數被加上了正規化函數，讓模型不能過於複雜，相當於限制模型擬和資料的能力
 原來 MSE: 25 ; Lasso MSE: 28.95

# **Ridge**

In [0]:
bo = datasets.load_boston()

In [0]:
# 切分訓練集/測試集
x2_train, x2_test, y2_train, y2_test = train_test_split(bo.data, bo.target, test_size= 0.2, random_state= 4)

In [0]:
# 建立一個線性回歸模型
ridge = linear_model.Ridge(alpha = 1.0)

In [0]:
# 將訓練資料丟進去模型訓練
ridge.fit(x2_train, y2_train)

#predict y
y2_pred = ridge.predict(x2_test)

In [27]:
# 印出各特徵對應的係數,可以很明顯看到比起 Linear Regression，參數的數值都明顯小了許多

ridge.coef_

array([-1.12499445e-01,  4.79562332e-02, -2.40438147e-02,  2.96199458e+00,
       -9.33966118e+00,  3.93079015e+00, -1.73821202e-02, -1.43347691e+00,
        2.75239392e-01, -1.38920708e-02, -8.31116943e-01,  9.15637729e-03,
       -4.66460539e-01])

In [28]:
regr.coef_

array([-1.15966452e-01,  4.71249231e-02,  8.25980146e-03,  3.23404531e+00,
       -1.66865890e+01,  3.88410651e+00, -1.08974442e-02, -1.54129540e+00,
        2.93208309e-01, -1.34059383e-02, -9.06296429e-01,  8.80823439e-03,
       -4.57723846e-01])

In [29]:
print("MSE: %.2d" %mean_squared_error(y2_test, y2_pred))

MSE: 25


In [30]:
print("regr MSE: %.2d" %mean_squared_error(y_test, y_pred))
print("lasso MSE: %.2d" %mean_squared_error(y1_test, y1_pred))

regr MSE: 25
lasso MSE: 28


可以看見 LASSO 與 Ridge 的結果並沒有比原本的線性回歸來得好